packages/authentication/test/unit/authentication.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from 'testlab';
+import {Authentication} from '../../src/authentication';
+import {BasicStrategy} from 'passport-http';
+const http = require('http');
+
+describe('Authentication', () => {
+  let passport;
+  let strategy;
+  let req;
+  let res;
+
+  describe('Basic Strategy', () => {
+    beforeEach(createBasicStrategy);
+    beforeEach(createPassport);
+    beforeEach(createMockHttpObjects);
+
+    it('authenticates successfully for correct credentials', () => {
+      const credential = 'simpson' + ':' + 'alpha';
+      const hash = new Buffer(credential).toString('base64');
+      req.headers.authorization = 'Basic ' + hash;
+      passport.authenticate(req, res, function done(err) {
+        expect(err).to.be.undefined;
+        expect(req.user.role).to.equal('admin');
+      });
+    });
+
+    it('returns error for invalid credentials', () => {
+      const credential = 'Flintstone' + ':' + 'invalid';
+      const hash = new Buffer(credential).toString('base64');
+      req.headers.authorization = 'Basic ' + hash;
+      passport.authenticate(req, res, function done(err) {
+        expect(err).to.equal('Incorrect password');
+      });
+    });
+
+    it('sets authentication header for reAuthentication', () => {
+      const credential = 'George1' + ':' + 'gamma';
+      const hash = new Buffer(credential).toString('base64');
+      req.headers.authorization = 'Basic ' + hash;
+      passport.authenticate(req, res);
+      expect(res.headers).to.have.property('WWW-Authenticate');
+    });
+  });
+
+  function createPassport() {
+    passport = new Authentication(strategy);
+  }
+
+  function createMockHttpObjects() {
+    req = {headers: {authorization: {}}, user: {role: {}}};
+    res = Object.create(http.ServerResponse.prototype);
```

Comment:
This is pretty dangerous, what is going to happen when somebody treats your mocked response as a writable stream (which it is) and calls `res.write` for example? Why is it necessary to inherit from `http.ServerResponse` at all?